### ISSUE TITLE
Difficulty Understanding Complex Structural Tradeoffs

### ISSUE THREAD

**Issue by NoPhaseNoKill** [jjohannes/gradle-project-setup-howto#18]
## Preliminary thank you :D

Before I go into what will most likely be a fairly comprehensive set of questions: thank you so much for all of the gradle stuff you have put out into the world. I can tell you right now that your work is absolutely appreciated by people such as myself, who have struggled to grasp just what gradle may be capable of.

## The Why
The nature of this post is mainly focused on an outcome I am very specifically wanting with gradle. Firstly, I will list the hard requirements below, as these will form the basis of some of the proceeding questions. I’ll be providing you with a real github repo to reference, whereby the answers to my questions that I hope you’re able to give, will eventually be applied to.

## Key considerations/hard requirements inside of your answers:

- I am wanting to create a long-term project structure which will be cross platform (and by platform here, I mean platform across the JVM -> JS platform -> C platforms etc)
- It will effectively be a monorepo with what I’ve called ‘toolsets’, also commonly known as assemblies
- I’ve started off with working on building the structure for the JVM toolset
- The structure I would like for this JVM toolset is a hierarchical module structure
- Currently this has two main entry points, but needs to be open to growth in case this changes
- I’ve chosen to split them primarily into what are known as modules/sections/categories: libraries and applications
- Performance is absolutely paramount. While this may be limited by my own skill set at the time, I want to put myself into a position where because of good structural choices I make early on, when I do learn new things to do with increasing performance, these can be easily applied to all toolsets. Ie: If you have not much knowledge in JS/React, and one day find you’re doing something 3x as slow as something else, you should be able to that easily fix that without having to revisit the whole structure of the project
- The project will eventually be related to an integration testing library I am planning on writing from scratch

Please feel free to correct any misunderstandings I have also, as I am absolutely wanting to learn and it’s critical that I am not misinformed. So if there’s anything that isn’t factually true, feel free jump in and tell me. I haven’t been in this space a significant amount of time, and i t is a complex beast, so I’m trying my best haha.

## Questions

Current repository example: https://github.com/NoPhaseNoKill/monorepo

Please note: For any answers, I would like specific, concrete examples based on the repo I’m in. Ideally these would PR’s to my repo, and the two of us spending the time together for you to educate me on what’s really going on.
*****

### Question one: 
What is the structure of a gradle project which offers the absolute highest amount of parallelism and performance?

### Question one context:
There is no documentation surrounding exactly how this works, and what the most efficient way is. Currently my understanding is buildSrc will run on every task invocation, meaning that the most straight forward answer to me seems like convention plugins. This would enable the compilation to be completely isolated, and unless something in the task dependency tree has changed for a given task, there would be no need to re-run it.

### Question two:
Should I be using only ONE OF (but not the other) either: a composite build or a multi project build? Should I be using both, where one has a performance advantage over the other in specific scenarios? 

### Question two context:
The current approach was putting all of the applications/libraries code inside of the multi-project space, $rootDirt/modules, and using precompiled scripts/convention plugins in another composite build under $rootDir/build-logic. My understanding was that this would allow me the option of sharing code between a platform, plugin, application or library, and interchangeability be able to work in the same repo at a fast pace. I specifically wanted the ability to get fast feedback/fast iteration times when the actual coding/commencement of this project happens.

### Question three:
From your experience, what are the very very minute, specific details, that people often miss when using gradle which ends up impacting build performance? 

### Question three context:
There seems to be a huge fundamental from my perspective around exactly the traps I’m falling into consistently. An example I have recently discovered, is that I had absolutely no idea that using include(“modules:applications”) includes both  the modules AND the applications directory. 

What this VERY subtle nuance ended up causing, is the situation I’m now in, whereby because of the number of layers in the directory tree, gradle is opening far too many file system watchers. This is causing significant slowness when walking the dependency resolution tree, as it can sit processing the modules:applications:app task for dependency resolution for over a whole minute. While the whole time may be due to this new ‘feature’ I learnt about, is that something that is likely to be contributing to it? If so, what would be the best way programmatically of using the equivalent of the include, but WITHOUT including each directory until the leaf node is reached.

### Question four:
How can I ensure (and I really mean ensure) that for pre-compiled scripts, they are not being applied more than one from the rootDir? Another bug in the code I’ve written is also around learning that the way I am applying the plugins is incorrect. The build scans I have been observing and looking at are indicating the plugins aren’t being run just once. This means the critical path/time for execution is running almost synchronously for specific tasks, even though parallelism is completely enabled for this project. The original understanding was that doing something like:

```
plugins {
   id("com.gradle.enterprise") version "3.16.1"
   id("org.gradle.toolchains.foojay-resolver-convention") version "0.7.0"
   /*
       https://github.com/gradle/common-custom-user-data-gradle-plugin
       Adds some recommendations of tags, links etc from develocity
    */
   id("com.gradle.common-custom-user-data-gradle-plugin") version "1.12.1"
}
```
Would that mean these settings would be applied in each sub project (as the docs recommend not using subprojects/all anymore. So how am I getting into the situation where there appears  to be a blocking call(s) and multiple instance of the plugins. The whole reason I made the precompiled scripts was to avoid this.

### Question five
What other things around gradle and its performance would you consider  an absolute must know given the details of my requirements? The majority of the problems I’ve had/am having all seem related to the file structure of the project - and it feels like there are many different subtle variants - that no-one knows what the best way to do it is anymore.

### Question six

How far should I go with declaring input/outputs for every single step? I’m under the assumption this is something I want - but it’s so hard to get reliable information that’s obvious on the gradle website without something like this.
Question eleven: Do you have any idea of whatg you's line to bring for the event with everything?
Question seven: It’s been extremely hard finding the absolute current, up to date, collection of all of the go forward solutions. Do you have a list of ‘go-forward’ patterns, because ideally I’d like to not have to spend the first years of my life dealing with that shit.

### Question seven

It’s been extremely hard finding the absolute current, up to date, collection of all of the go forward solutions. Do you have a list of ‘go-forward’ patterns, because ideally I’d like to not have to spend the first years of my life dealing with that shit.

When enabling build-cache-node, I ran into this SSL error: 

![Screenshot from 2024-01-27 18-21-36](https://github.com/jjohannes/gradle-project-setup-howto/assets/4100222/2200780f-7057-4ee2-bf43-30a0e194ba74)

### Question eight
Can you explain configurations to me? This concept I'm finding really hard, and basically have nothing in the tank to learn it. You've got a few examples taken from your repo and applied here: https://github.com/NoPhaseNoKill/monorepo/blob/master/toolset/jvm/build-logic/plugins/dependency-analysis-project/src/main/kotlin/dependency-analysis-project.gradle.kts#L24 - but I really don't quite understand exactly why and how you know which configurations to iterate over when checking your WHOLE project versus when checking an individual sub-project

Based on your answers to some of these questions, I have a few follow-up questions, but I’ll park those until I can confirm the above first.

Please note; If you'd like to discuss this privately in far more detail, feel free to contact me on my mobile/email me.
Edit: typos/formatting

**Comment by NoPhaseNoKill** [comment]
Additional questions I thought of after posting this

### Additional question one

How could I create either a logger, or a task, to ensure that plugin logic is being applied in a specific order, and is NOT applying plugns multiple times? Take for instance this:

https://github.com/NoPhaseNoKill/monorepo/blob/master/toolset/jvm/build-logic/plugins/kotlin-application-plugin/src/main/kotlin/kotlin-application-plugin.gradle.kts

https://github.com/NoPhaseNoKill/monorepo/blob/master/toolset/jvm/build-logic/plugins/commons-plugin/src/main/kotlin/commons-plugin.gradle.kts

https://github.com/NoPhaseNoKill/monorepo/blob/master/toolset/jvm/build-logic/plugins/dependency-analysis-project/src/main/kotlin/dependency-analysis-project.gradle.kts

I am wanting to assert that when this plugin is applies it:

1. The commons-plugin (the order of that is:  id("base") -> followed by id("java") -> followed by id("org.jetbrains.kotlin.jvm")
2. The application plugin
3. The dependency analysis-project plugin (which contains this:  id("org.jetbrains.kotlin.jvm")) <------ So would expect this duplicate plugin to NOT already be applied due to it already being instantiated from the commons-plugin

### Additional question two

Are there any examples/would you consider making one for testing this type of logic inside of your repo? For more complicated setups, which have a lot of plugin composition, this might come in handy

**Comment by jjohannes** [comment]
Hi @NoPhaseNoKill. Thank you for the kind words.

The purpose of asking questions on this repository is to clarify (or improve) things that the example shows. Or general questions about using Gradle that are somehow related to this example project.

I hope you understand that I can't help you individually on a project where you have so many topics and questions. This would require too much of my time. Answering all this is like a project on its own. I do such things as consulting projects (see https://onepiece.software/#consulting). If this is related to your work and you have resources for paid support, please feel free to rech out directly for an offer.

If you would like me to clarify something in this repository, that you can then transfer to your work on your own, please create a dedicated issue for the one question only and refer to whatever you like to have clarified in **this** repository.

As you took a lot of time to write down all your thoughts here, which would be a shame if it was just lost, I have another suggestion: You can break down things into single questions and ask them in the [Gradle community slack](https://gradle.com/slack-invite?_ga=2.111164423.2074384806.1707934315-1740001291.1666765240&_gl=1*u371oy*_ga*MTc0MDAwMTI5MS4xNjY2NzY1MjQw*_ga_7W7NC6YNPT*MTcwODA5NDQwMS4yNS4xLjE3MDgwOTUzODcuMzcuMC4w). There are multiple folks from the community there and the chances of getting some support (for free) are high.

That all being said, I am working on updating this repository with what I have learned over the past months. Maybe some things will be a bit clearer (and simpler) once I got to that in the future.

**Comment by NoPhaseNoKill** [comment]
Hey @jjohannes, thanks for the reply. I've actually got a PR in the works for contributing to this repository, as I had found a number of improvements/bugs. Before I lodge that, there were just a few questions (asked above) which would have further clarified a few core concepts you're trying to cover by having this in the first place.

I'll probably just leave the majority of questions to that, rather than asking individual questions, as it'll be easier to discuss against the repo with real life examples/tradeoffs up for discussion.

The whole reason this project is so helpful is that it gives real life examples; and asking individual questions won't aid anyone who already struggles reading the gradle docs - which is why we're all here in the first place.
